### PR TITLE
Project: Guess the Word

### PR DESCRIPTION
**Pull request by Kaktug**
# Description

This PR includes my solution to The Odin Project's "Guess the Word" assignment.   The prompt for the assignment can be seen here:
https://www.theodinproject.com/lessons/ruby-hangman

> ## Assignment
> You will be building a simple command line Hangman game where one player plays against the computer, but a bit more advanced.  If you're unfamiliar with how Hangman works, see [Wikipedia](https://en.wikipedia.org/wiki/Hangman_(game)).
>
> 1. Download the `google-10000-english-no-swears.txt` dictionary file from the first20hours GitHub repository [google-10000-english](https://raw.githubusercontent.com/first20hours/google-10000-english/master/google-10000-english-no-swears.txt)
> 2. When a new game is started, your script should load in the dictionary and randomly select a word between 5 and 12 characters long for the secret word.
> 3. You don't need to draw an actual stick figure (though you can if you want to!), but do display some sort of count so the player knows how many more incorrect guesses they have before the game ends.  You should also display which correct letters have already been chosen (and their position in the word, e.g. `_ r o g r a _ _ i n g`) and which incorrect letters have already been chosen.
> 4. Every turn, allow the player to make a guess of a letter.  It should be case insensitive.  Update the display to reflect whether the letter was correct or incorrect.  If out of guesses, the player should lose.
> 5. Now implement functionality where, at the start of any turn, instead of making a guess the player should also have the option to save the game.  Remember what you learned about serializing objects... you can serialize your game class too!
> 6. When the program first loads, add in an option that allows you to open one of your saved games, which should jump you exactly back to where you were when you saved.  Play on!

# Screenshots

### New Game:
Screen Shot 2023-04-10 at 14 28 56

### Playing the game:
_Note the incorrect guesses, correct guesses, and the shark's progress towards the swimmer_
Screen Shot 2023-04-10 at 14 29 09

### Saving the game:
Screen Shot 2023-04-10 at 14 29 19

### Recalling a saved game:
_prompt to load a detected save file:_
Screen Shot 2023-04-10 at 14 29 39

_the restored game, via the save file data:_
Screen Shot 2023-04-10 at 14 29 48

### Losing the game:
Screen Shot 2023-04-10 at 14 31 04

### Winning the game:
Screen Shot 2023-04-10 at 14 30 30

# Notes / Thoughts

1. I wasn't super comfortable with the "Hangman" imagery so I spun the assignment into a Shark Attack-based one.  Hopefully that is OK (I think it still hits all the major gameplay elements)
2. The `GameFile` class feels a bit messy
3. I'm not sure if modules are the best way to implement the `Displayable` and `Wordable` features.  Originally I had set them up as classes but that made it really hard for me to figure out serialization of the data.  Some of it was owned by each of the three classes so I couldn't make sense of saving/recalling all those properties.  If modules are usually used in a "Has-a" relationship then I think my refactor ended up OK (e.g. Game HAS A display and HAS A word to guess).

### CODE REVIEW

File: full-stack-ruby-on-rails/project-guess-the-word/lib/displayable.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+# A module which helps to show the player feedback as the game progresses
+module Displayable
+  WAVE_TOP = '.--.'
+  WAVE_SPACER = ' '
+  SHARK_FIN = '𓂄'
+  SWIMMER = '.º¬'
+  WAVE_COUNT = 9
+  # 𓂄 .--.\s
```
Comment (chmurph2) [inline full-stack-ruby-on-rails/project-guess-the-word/lib/displayable.rb:10]: Good use of inline comments.

**Review by chmurph2 — COMMENTED** [review]
> 1. I wasn't super comfortable with the "Hangman" imagery so I spun the assignment into a Shark Attack-based one. Hopefully that is OK (I think it still hits all the major gameplay elements)

Yeah, good call, though neither way is a good way to go. 😅

> 2. The GameFile class feels a bit messy

Seemed OK to me.

> 3. I'm not sure if modules are the best way to implement the Displayable and Wordable features. Originally I had set them up as classes but that made it really hard for me to figure out serialization of the data. Some of it was owned by each of the three classes so I couldn't make sense of saving/recalling all those properties. If modules are usually used in a "Has-a" relationship then I think my refactor ended up OK (e.g. Game HAS A display and HAS A word to guess).

I left feedback on this one.

File: full-stack-ruby-on-rails/project-guess-the-word/shark_attack_game.rb [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+require './lib/wordable'
+require './lib/displayable'
+require './lib/game_file'
+
+# A class to manage our game's state
+class SharkAttackGame
+  include Wordable
+  include Displayable
+
+  def initialize
```
[the remark this answers is not in the document]
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-guess-the-word/shark_attack_game.rb:12]: Thanks, Chris.  Now that I've got a slightly better handle on serialization I'll work on refactoring those two modules into classes today.

File: full-stack-ruby-on-rails/project-guess-the-word/shark_attack_game.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  end
+
+  def load_game?
+    return false unless GameFile.save_exists?
+
+    puts 'Load saved game? (y/n)'
+    yes_response?
+  end
+
+  def load
```
[the remark this answers is not in the document]
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-guess-the-word/shark_attack_game.rb:56]: :exploding_head: yes, that makes sense to me

File: full-stack-ruby-on-rails/project-guess-the-word/lib/game_file.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    ensure_save_directory_exists
+
+    save_file = File.open(@save_path, 'w')
+    save_file.puts YAML.dump(@data)
+    save_file.close
+    puts 'Game saved.'
+  end
+
+  def ensure_save_directory_exists
+    Dir.mkdir(@save_directory) unless Dir.exist?(@save_directory)
```
[the remark this answers is not in the document]
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-guess-the-word/lib/game_file.rb:41]: Works great, thank you :bow:

File: full-stack-ruby-on-rails/project-guess-the-word/lib/displayable.rb (hunk 1 above)
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-guess-the-word/lib/displayable.rb]: :bow:

File: full-stack-ruby-on-rails/project-guess-the-word/lib/display.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  SHARK_FIN = '𓂄'
+  SWIMMER = '.º¬'
+  WAVE_COUNT = 9
+  # 𓂄 .--.\s
+  SHARK_LINE = "#{SHARK_FIN}#{WAVE_SPACER}#{WAVE_TOP}#{WAVE_SPACER}"
+  # .--.\s
+  EMPTY_LINE = "#{WAVE_TOP}#{WAVE_SPACER}"
+
+  def update(word, correct_guesses, incorrect_guesses, game_won: nil)
+    game_won.nil? ? draw_round(word, correct_guesses, incorrect_guesses) : draw_game_over(word, game_won)
```
Comment (chmurph2) [inline full-stack-ruby-on-rails/project-guess-the-word/lib/display.rb:16]: nit: I recommend using an if/else condition for this. It's easier to scan over and reason about vs. long ternaries that stretch off to the horizon.

File: full-stack-ruby-on-rails/project-guess-the-word/shark_attack_game.rb [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+require './lib/game_file'
+
+# A class to manage our game's state
+class SharkAttackGame
+  def initialize(dictionary = Dictionary.new, display = Display.new, save_file = GameFile.new)
+    @save_file = save_file
+    @correct_guesses = []
+    @incorrect_guesses = []
+    @word_to_guess = load_game? ? load! : dictionary.random_word
+    @display = display
```
Comment (chmurph2) [inline full-stack-ruby-on-rails/project-guess-the-word/shark_attack_game.rb:14]: very minor: I think an unspoken convention is to set any params passed in to their ivars, at the top of the initializer.

**Pull request merged**